### PR TITLE
Add Bounding Box to Field Line Integration

### PR DESCRIPTION
**Pull request by dpanici**
- Adds a bounding box to the `field_line_integrate` defined by `bounds_R` and `bounds_Z` keyword arguments, which form a hollow cylindrical bounding box. If the field line trajectory exits these bounds, the RHS will be multiplied by `exp(-(R**2+Z**2))` to stop the trajectory and prevent tracing the field line out to infinity, which is both costly and unnecessary when making a Poincare plot, which is the principle purpose of the function.

In one example I had when tracing the precise_QA with a coilset that had some diverging field lines, it took 1/3 of the time compared to without using bounds, so seems to work fine. Defaults to no bounds.

@ddudt this PR is not addressing anything that happens when a SplineMagneticField is used and the field lines exit its domain of validity when `extrap=False` unless you specify the bounding box to lie within its domain of validity, in which case it should stop the integration soon after it exits the bounding box and possibly before it exits the field's interpolation domain. If there are ideas of how to treat that then we can make an issue and a separate PR

### CODE REVIEW

File: desc/magnetic_fields.py [hunk 1]
```diff
@@ -1419,12 +1436,35 @@ def field_line_integrate(
...
+                jnp.array(
+                    [
+                        jnp.less(r, bounds_R[0]),
+                        jnp.greater(r, bounds_R[1]),
+                        jnp.less(z, bounds_Z[0]),
+                        jnp.greater(z, bounds_Z[1]),
+                    ]
+                )
+            ),
+            jnp.exp(-(r**2 + z**2)),
```
[the remark this answers is not in the document]
Comment (dpanici) [reply desc/magnetic_fields.py]: I think what I have should be continuous now, the decay factor is now `exp((r-R0)**2) * exp((r-R1)**2) * exp((z-Z0)**2) * exp((z-Z1)**2) `

where each element of this product is only added if that constraint is violated, and the bounding box is $R \in [R0,R1], Z \in [Z0,Z1]$

I also added a `1e-2` factor outside of the exponential to accelerate the decay a bit, as without it the integration still goes on for a bit longer (made it hard to test that this was working)

File: desc/magnetic_fields.py [hunk 2]
```diff
@@ -1419,12 +1436,46 @@ def field_line_integrate(
...
+                    jnp.exp(-((r - bounds_R[1]) ** 2)),
+                    jnp.exp(-((z - bounds_Z[0]) ** 2)),
+                    jnp.exp(-((z - bounds_Z[1]) ** 2)),
+                ]
+            ),
+            1.0,
+        )
+        # multiply all together, the conditions that are not violated
+        # are just one while the violated ones are continous decaying exponentials
+        decay_factor = jnp.prod(decay_factor)
```
Comment (f0uriest) [inline desc/magnetic_fields.py:1468]: does this need to be along an axis? Will this affect all field lines or just the one that leaves?

File: desc/magnetic_fields.py [hunk 3]
```diff
@@ -1419,12 +1436,46 @@ def field_line_integrate(
...
+        decay_factor = jnp.where(
+            jnp.array(
+                [
+                    jnp.less(r, bounds_R[0]),
+                    jnp.greater(r, bounds_R[1]),
+                    jnp.less(z, bounds_Z[0]),
+                    jnp.greater(z, bounds_Z[1]),
+                ]
+            ),
+            1e-2
```
Comment (f0uriest) [inline desc/magnetic_fields.py:1453]: I think this makes it discontinuous as well. The decay factor should be 1 for points on the boundary. If we want it to decay quicker probably better to just weight the exponential

Comment (dpanici) [reply desc/magnetic_fields.py:1453]: dang you are right, thanks

File: tests/test_magnetic_fields.py [hunk 4]
```diff
@@ -678,6 +678,18 @@ def test_field_line_integrate(self):
...
         np.testing.assert_allclose(r[-1], 10, rtol=1e-6, atol=1e-6)
         np.testing.assert_allclose(z[-1], 0.001, rtol=1e-6, atol=1e-6)
+        # test that bounds work correctly, and stop integration when trajectory
+        # hits the bounds
+        r0 = [10.1]
+        z0 = [0.0]
+        # this will hit the R bound
+        # (there is no Z bound, and R would go to 10.0 if not bounded)
+        r, z = field_line_integrate(r0, z0, phis, field, bounds_R=(10.05, np.inf))
+        np.testing.assert_allclose(r[-1], 10.05, rtol=3e-4)
```
Comment (dpanici) [inline tests/test_magnetic_fields.py:688]: we don't expect the integration to stop immediately after the bound is exited, but it should stop soon after depending on how fast the RHS decays. I made the decay very fast so we can test it, but if for some reason we should lessen the decay rates, then we will need to make these test tols here less tight

**Review by f0uriest — DISMISSED** [review]

File: desc/magnetic_fields.py [hunk 5]
```diff
@@ -1419,12 +1436,45 @@ def field_line_integrate(
...
+                ]
+            ),
+            jnp.array(
+                [
+                    # we mult by 1e6 to accelerate the decay so that the integration
+                    # is stopped soon after the bounds are exited.
+                    jnp.exp(-(1e6 * (r - bounds_R[0]) ** 2)),
+                    jnp.exp(-(1e6 * (r - bounds_R[1]) ** 2)),
+                    jnp.exp(-(1e6 * (z - bounds_Z[0]) ** 2)),
+                    jnp.exp(-(1e6 * (z - bounds_Z[1]) ** 2)),
```
Comment (ddudt) [inline desc/magnetic_fields.py:1460]: Can we make this decay rate an input argument? A default value of 1e6 is fine, but that way it isn't a hard-coded magic number

File: desc/magnetic_fields.py [hunk 6]
```diff
@@ -1401,6 +1403,21 @@ def field_line_integrate(
         relative and absolute tolerances for ode integration
     maxstep : int
         maximum number of steps between different phis
+    bounds_R : tuple of (float,float), optional
+        R bounds for field line integration bounding box.
+        If supplied, the RHS of the field line equations will be
+        multipled by exp(-r) where r is the distance from the origin,
```
Comment (ddudt) [inline desc/magnetic_fields.py:1409]: where r is the distance from the bounding box

File: desc/magnetic_fields.py [hunk 7]
```diff
@@ -1401,6 +1403,21 @@ def field_line_integrate(
...
+        this is meant to prevent the field lines which escape to infinity from
+        slowing the integration down by being traced to infinity.
+        defaults to (0,np.inf)
+    bounds_Z : tuple of (float,float), optional
+        Z bounds for field line integration bounding box.
+        If supplied, the RHS of the field line equations will be
+        multipled by exp(-r) where r is the distance from the origin,
```
Comment (ddudt) [inline desc/magnetic_fields.py:1416]: where r is the distance from the bounding box

File: tests/test_magnetic_fields.py [hunk 8]
```diff
@@ -678,6 +678,18 @@ def test_field_line_integrate(self):
         r, z = field_line_integrate(r0, z0, phis, field)
         np.testing.assert_allclose(r[-1], 10, rtol=1e-6, atol=1e-6)
         np.testing.assert_allclose(z[-1], 0.001, rtol=1e-6, atol=1e-6)
+        # test that bounds work correctly, and stop integration when trajectory
+        # hits the bounds
```
Comment (ddudt) [inline tests/test_magnetic_fields.py:682]: Maybe make this a new separate unit test instead of adding it onto this existing test?

**Review by ddudt — CHANGES_REQUESTED** [review]

File: desc/magnetic_fields.py (hunk 5 above)
Comment (dpanici) [reply desc/magnetic_fields.py:1460]: good point, I'll add

**Review by f0uriest — APPROVED** [review]

File: tests/test_magnetic_fields.py (hunk 8 above)
Comment (ddudt) [reply tests/test_magnetic_fields.py:682]: This is duplicated now that you copied it to a new unit test, so you can delete it.

**Review by ddudt — CHANGES_REQUESTED** [review]

**Review by ddudt — APPROVED** [review]
